iap/iap.go

```diff
@@ -0,0 +1,515 @@
+	var wrapper map[string]interface{}
+	if err := json.Unmarshal([]byte(receipt), &wrapper); err != nil {
+		return nil, err
+	}
+
+	unwrapped, ok := wrapper["json"].(string)
+	if !ok {
+		return nil, errors.New("'json' field not found, receipt is malformed")
+	}
+
+	var gr ReceiptGoogle
+	if err := json.Unmarshal([]byte(unwrapped), &gr); err != nil {
+		return nil, err
+	}
+	return &gr, nil
+}
+
+// Request an authenticated context (token) from Google for the Android publisher service.
+func GetGoogleAccessToken(ctx context.Context, httpc *http.Client, email string, privateKey string) (string, error) {
+	const authUrl = "https://accounts.google.com/o/oauth2/token"
+
+	type GoogleClaims struct {
+		Scope string `json:"scope,omitempty"`
+		jwt.StandardClaims
+	}
+
+	claims := &GoogleClaims{
+		"https://www.googleapis.com/auth/androidpublisher",
+		jwt.StandardClaims{
+			Audience:  authUrl,
+			ExpiresAt: time.Now().Add(1 * time.Hour).Unix(),
+			IssuedAt:  time.Now().Unix(),
+			Issuer:    email,
+		},
+	}
+
+	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+	block, _ := pem.Decode([]byte(privateKey))
+	if block == nil {
+		return "", errors.New("pem.Decode error")
+	}
+
+	pk, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err != nil {
+		return "", err
+	}
+	signed, err := token.SignedString(pk)
+	if err != nil {
+		return "", err
+	}
+
+	data := url.Values{}
+	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer")
+	data.Set("assertion", signed)
+	body := data.Encode()
+
+	req, err := http.NewRequestWithContext(ctx, "POST", authUrl, strings.NewReader(body))
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := httpc.Do(req)
+	if err != nil {
+		return "", err
+	}
+	//goland:noinspection GoUnhandledErrorResult
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case 200:
+		buf, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+
+		var out map[string]interface{}
+		if err := json.Unmarshal(buf, &out); err != nil {
+			return "", err
+		}
+		return out["access_token"].(string), nil
+	default:
+		return "", fmt.Errorf("non-200 response from Google auth: %+v", resp)
+	}
+}
+
+// Validate an IAP receipt with the Android Publisher API and the Google credentials.
+func ValidateReceiptGoogle(ctx context.Context, httpc *http.Client, clientEmail string, privateKey string, receipt string) (*ValidateReceiptGoogleResponse, *ReceiptGoogle, []byte, error) {
+	if len(clientEmail) < 1 {
+		return nil, nil, []byte{}, errors.New("'clientEmail' must not be empty")
+	}
+
+	if len(privateKey) < 1 {
+		return nil, nil, []byte{}, errors.New("'privateKey' must not be empty")
+	}
+
+	if len(receipt) < 1 {
+		return nil, nil, []byte{}, errors.New("'receipt' must not be empty")
+	}
+
+	token, err := GetGoogleAccessToken(ctx, httpc, clientEmail, privateKey)
```

Comment:
Should a fresh access token be requested each time? If I recall last time I worked with similar Google services it should be possible to cache access tokens for some time.